src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/PrecodeStubs_1.cs

```diff
@@ -0,0 +1,183 @@
+        }
+    }
+
+    internal sealed class PInvokeImportPrecode : StubPrecode
+    {
+        internal PInvokeImportPrecode(TargetPointer instrPointer) : base(instrPointer, KnownPrecodeType.PInvokeImport) { }
+    }
+
+    internal sealed class FixupPrecode : ValidPrecode
+    {
+        internal FixupPrecode(TargetPointer instrPointer) : base(instrPointer, KnownPrecodeType.Fixup) { }
+        internal override TargetPointer GetMethodDesc(Target target, Data.PrecodeMachineDescriptor precodeMachineDescriptor)
+        {
+            TargetPointer fixupPrecodeDataAddress = InstrPointer + precodeMachineDescriptor.StubCodePageSize;
+            Data.FixupPrecodeData fixupPrecodeData = target.ProcessedData.GetOrAdd<Data.FixupPrecodeData>(fixupPrecodeDataAddress);
+            return fixupPrecodeData.MethodDesc;
+
+        }
+    }
+
+    internal sealed class ThisPtrRetBufPrecode : ValidPrecode // FIXME: is this a StubPrecode?
+    {
+        internal ThisPtrRetBufPrecode(TargetPointer instrPointer) : base(instrPointer, KnownPrecodeType.ThisPtrRetBuf) { }
+
+        internal override TargetPointer GetMethodDesc(Target target, Data.PrecodeMachineDescriptor precodeMachineDescriptor)
+        {
+            throw new NotImplementedException(); // TODO(cdac)
+        }
+    }
+
+    private bool IsAlignedInstrPointer(TargetPointer instrPointer) => _target.IsAlignedToPointerSize(instrPointer);
+
+    private byte ReadPrecodeType(TargetPointer instrPointer)
+    {
+        if (MachineDescriptor.ReadWidthOfPrecodeType == 1)
+        {
+            byte precodeType = _target.Read<byte>(instrPointer + MachineDescriptor.OffsetOfPrecodeType);
+            return (byte)(precodeType >> MachineDescriptor.ShiftOfPrecodeType);
+        }
+        else if (MachineDescriptor.ReadWidthOfPrecodeType == 2)
+        {
+            ushort precodeType = _target.Read<ushort>(instrPointer + MachineDescriptor.OffsetOfPrecodeType);
+            return (byte)(precodeType >> MachineDescriptor.ShiftOfPrecodeType);
+        }
+        else
+        {
+            throw new InvalidOperationException($"Invalid precode type width {MachineDescriptor.ReadWidthOfPrecodeType}");
+        }
+    }
+
+    private Data.StubPrecodeData GetStubPrecodeData(TargetPointer stubInstrPointer)
+    {
+        TargetPointer stubPrecodeDataAddress = stubInstrPointer + MachineDescriptor.StubCodePageSize;
+        return _target.ProcessedData.GetOrAdd<Data.StubPrecodeData>(stubPrecodeDataAddress);
+    }
+
+    private KnownPrecodeType? TryGetKnownPrecodeType(TargetPointer instrAddress)
+    {
+        // precode.h Precode::GetType()
+        byte precodeType = ReadPrecodeType(instrAddress);
+        if (precodeType == MachineDescriptor.StubPrecodeType)
+        {
+            // get the actual type from the StubPrecodeData
+            Data.StubPrecodeData stubPrecodeData = GetStubPrecodeData(instrAddress);
+            precodeType = stubPrecodeData.Type;
+        }
```

Comment:
@elinor-fung the suggestion behaves differently.  For some kinds of precode, the precode type returned from `instrAddress` is just an approximation - it looks like a stub, but the StubPrecodeData gives a more precise precode type (for example, Fixup).  So this is a two-phase operation:
1. Read the approximate precode type.  
2. If it's "stub", read the exact precode type from the stub data (it might still be "stub", but it could also be "fixup" or something else).